tls/s2n_cipher_suites.c

```diff
@@ -1168,11 +1171,164 @@ static int s2n_wire_ciphers_contain(const uint8_t *match, const uint8_t *wire, u
+    RESULT_ENSURE_REF(chacha20_cipher);
+
+    *chacha20_cipher = NULL;
+
+    for (size_t i = 0; i < security_policy->cipher_preferences->count; i++) {
+        struct s2n_cipher_suite *cipher = security_policy->cipher_preferences->suites[i];
+        const uint8_t *server_iana = cipher->iana_value;
+
+        bool cipher_uses_chacha20_alg = false;
+        RESULT_GUARD(s2n_cipher_suite_uses_chacha20_alg(cipher, &cipher_uses_chacha20_alg));
+
+        if (cipher_uses_chacha20_alg && s2n_wire_ciphers_contain(server_iana, wire_stuffer, count, cipher_suite_len)) {
+            *chacha20_cipher = cipher;
+            return S2N_RESULT_OK;
+        }
+    }
+
+    return S2N_RESULT_OK;
+}
+
+/* If both server and client signaled support for ChaCha20 boosting, then the corresponding ChaCha20 cipher suite is set. Else, NULL is set. */
+static S2N_RESULT s2n_get_boosted_cipher_suite(struct s2n_connection *conn, struct s2n_stuffer wire_stuffer, uint32_t count, uint32_t cipher_suite_len, struct s2n_cipher_suite **boosted_cipher_suite) {
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(boosted_cipher_suite);
+
+    const struct s2n_security_policy *security_policy;
+    RESULT_GUARD_POSIX(s2n_connection_get_security_policy(conn, &security_policy));
+
+    /* First check if the server has enabled chacha20 boosting */
+    *boosted_cipher_suite = NULL;
+    if (!security_policy->cipher_preferences->allow_chacha20_boosting) {
+        return S2N_RESULT_OK;
+    }
+
+    /* Next, check if the client has a chacha20 cipher as its most preferred cipher suite */
+    uint8_t clients_first_cipher_iana[S2N_SSLv2_CIPHER_SUITE_LEN] = { 0 };
+    RESULT_GUARD_POSIX(s2n_stuffer_read_bytes(&wire_stuffer, clients_first_cipher_iana, cipher_suite_len));
+
+    struct s2n_cipher_suite *client_first_cipher_suite = NULL;
+    RESULT_GUARD(s2n_cipher_suite_from_iana(clients_first_cipher_iana, &client_first_cipher_suite));
+    RESULT_ENSURE_REF(client_first_cipher_suite);
+
+    bool cipher_suite_has_chacha20_alg = false;
+    RESULT_GUARD(s2n_cipher_suite_uses_chacha20_alg(client_first_cipher_suite, &cipher_suite_has_chacha20_alg));
+    if (!cipher_suite_has_chacha20_alg) { 
+        return S2N_RESULT_OK;
+    }
+
+    /* The client has signaled ChaCha20 boosting. Now, iterate through the server ciphersuite preferences for a negotiable ChaCha20 cipher */
+    struct s2n_cipher_suite *negotiated_chacha20_cipher_suite = NULL;
+    RESULT_GUARD(s2n_negotitate_chacha20_cipher_suite(security_policy, count, cipher_suite_len, wire_stuffer, &negotiated_chacha20_cipher_suite));
+    if (!negotiated_chacha20_cipher_suite) {
+        return S2N_RESULT_OK;
+    }
+
+    /* Validate that the chacha20 cipher suite is a valid for the connection */
+    bool chacha20_is_valid_match = false;
+    RESULT_GUARD(s2n_validate_cipher_suite_match(conn, &negotiated_chacha20_cipher_suite, &chacha20_is_valid_match));
+    if (!chacha20_is_valid_match) {
+        return S2N_RESULT_OK;
+    }
```

Comment:
You need to choose the first valid chacha20 cipher suite. You can't give up if the first match isn't valid. That could just mean we only have an RSA cert and the first match requires an ECDSA cert, which amounts to the same behavior as [my previous example](https://github.com/aws/s2n-tls/pull/3543#discussion_r998687306).

Reply:
Right, I'll move the validation to the `s2n_negotitate_chacha20_cipher_suite` method.